oauthproxy.go

```diff
@@ -571,22 +571,37 @@ func (p *OAuthProxy) ManualSignIn(req *http.Request) (string, bool) {
 func (p *OAuthProxy) GetRedirect(req *http.Request) (redirect string, err error) {
 	err = req.ParseForm()
 	if err != nil {
-		return
+		return redirect, err
 	}
 
 	redirect = req.Header.Get("X-Auth-Request-Redirect")
 	if req.Form.Get("rd") != "" {
 		redirect = req.Form.Get("rd")
 	}
 	if !p.IsValidRedirect(redirect) {
+		proto := ""
+		host := ""
 		// Use RequestURI to preserve ?query
-		redirect = req.URL.RequestURI()
-		if strings.HasPrefix(redirect, p.ProxyPrefix) {
-			redirect = "/"
+		uri := req.URL.RequestURI()
+
+		// Quirk: On reverse proxies that doesn't have support for
+		// "X-Auth-Request-Redirect" header or dynamic header/query string
+		// manipulation (like Traefik v1 and v2), we can try if the header
+		// X-Forwarded-Host exists or not.
+		if p.IsForwardedRequest(req) {
+			proto = fmt.Sprintf("%s://", util.GetRequestProto(req))
+			host = util.GetRequestHost(req)
+			uri = util.GetRequestURI(req)
+		}
+
+		if strings.HasPrefix(uri, p.ProxyPrefix) {
+			uri = "/"
 		}
+
+		redirect = fmt.Sprintf("%s%s%s", proto, host, uri)
```

Comment:
Could this logic perhaps be refactored into a sub method that makes it easier to read through the code? Perhaps `getRedirectFromForwardedHeaders` or something? Then we could also test it separately as well which might be beneficial

Reply:
`getRedirectFromForwardHeaders` is a thing now.